File: src/ext/sample.py
import yaze

class YazePyExtension:
    def __init__(self):
        self.name = "My Python Extension"
        self.version = "1.0"

    def initialize(self, context):
        print(f"{self.name} Initialized with context: {context}")
        self.context = context
        self.register_event_hooks()

    def cleanup(self):
        print(f"{self.name} Cleaned Up")

    def manipulate_rom(self, rom):
        if rom and rom.data:
            print(f"First byte of ROM: 0x{rom.data[0]:02X}")
        else:
            print("ROM data is not loaded.")

    def extend_ui(self, context):
        import imgui
        if imgui.begin("My Python Extension Window"):
            imgui.text("Hello from My Python Extension!")
            imgui.end()

    def register_commands(self):
        # Register custom commands here
        print(f"{self.name} Commands Registered")

    def register_custom_tools(self):
        # Register custom tools here
        print(f"{self.name} Custom Tools Registered")

    def on_rom_loaded(self):
        print("ROM has been loaded!")

    def register_event_hooks(self):
        # Register event hooks, like for ROM loaded
        self.context.register_event_hooks(
            YAZE_EVENT_ROM_LOADED, self.on_rom_loaded)


def get_yaze_extension():
    return YazePyExtension()

File: src/ext/sample.c
#include <stdio.h>

#include "ext/extension.h"
#include "yaze.h"

void my_extension_initialize(yaze_editor_context* context) {
  printf("My Extension Initialized\n");
}

void my_extension_cleanup(void) { printf("My Extension Cleaned Up\n"); }

void my_extension_extend_ui(yaze_editor_context* context) {
  // Add a custom tab or panel to the editor
}

void my_extension_manipulate_rom(z3_rom* rom) {
  // Modify ROM data
}

void my_extension_register_commands(void) {
  // Register custom commands
}

void my_extension_handle_file_format(void) {
  // Handle custom file formats
}

void my_extension_register_event_hooks(yaze_event_type event,
                                       yaze_event_hook_func hook) {
  // Register event hooks for specific events
}

void my_extension_register_custom_tools(void) {
  // Register custom tools or editors
}

yaze_extension* get_yaze_extension(void) {
  static yaze_extension ext = {"My Extension",
                               "1.0",
                               my_extension_initialize,
                               my_extension_cleanup,
                               my_extension_manipulate_rom,
                               my_extension_extend_ui,
                               my_extension_register_commands,
                               my_extension_register_custom_tools,
                               my_extension_register_event_hooks};
  return &ext;
}